### PR TITLE
CORDA-540: Ensure that covariance of type is handled correctly when serializing with AMQP

### PR DESCRIPTION
**Pull request by vkolomeyko**

### CODE REVIEW

File: node-api/src/main/kotlin/net/corda/nodeapi/internal/serialization/amqp/DeserializationInput.kt [hunk 1]
```diff
@@ -142,10 +143,16 @@ class DeserializationInput(internal val serializerFactory: SerializerFactory) {
...
+     *  * [java.util.List<Command<?>>] and [java.lang.Class<? extends net.corda.core.contracts.Contract>]
+     *  * [T : Parent] and [Parent]
+     *
      * In the future tighter control might be needed
      */
     private fun Type.materiallyEquivalentTo(that: Type): Boolean =
-            asClass() == that.asClass() && that is ParameterizedType
+        when(that) {
+            is ParameterizedType -> asClass() == that.asClass()
+            is TypeVariable<*> -> isSubClassOf(that.bounds.single())
```
Comment (rick-r3) [inline node-api/src/main/kotlin/net/corda/nodeapi/internal/serialization/amqp/DeserializationInput.kt]: `single()` will error if there are two bounds.  Is that what we want?  Or do we just return false?  e.g.

`isSubClassOf(that.bounds.first()) && that.bound.size == 1`

Comment (vkolomeyko) [reply node-api/src/main/kotlin/net/corda/nodeapi/internal/serialization/amqp/DeserializationInput.kt]: Agree - will change to `first()`

File: node-api/src/test/java/net/corda/nodeapi/internal/serialization/amqp/ListsSerializationJavaTest.java [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+        }
+
+        public void setValue(int value) {
+            this.value = value;
+        }
+    }
+
+    @CordaSerializable
+    public static class CovariantContainer<T extends Parent> {
+        private List<T> content;
```
Comment (rick-r3) [inline node-api/src/test/java/net/corda/nodeapi/internal/serialization/amqp/ListsSerializationJavaTest.java]: Should this not have a constructor which takes content?

Comment (vkolomeyko) [reply node-api/src/test/java/net/corda/nodeapi/internal/serialization/amqp/ListsSerializationJavaTest.java]: Yes, I have tried few options in Java including having constructor, doesn't seem to work. Will continue today.

**Pull request merged**